README.md
# PureGuestbook
Pure book - pure project


# Following technologies is used:
bootstrap-4.3.1

fzaninotto/faker ^1.8

aura/sqlquery ^2.7

twig/twig ^2.0

history.txt
24.04.19 - New project

index.php
<?php
require_once __DIR__ . '/vendor/autoload.php';
require_once __DIR__ . '/functions/PDOconn.php';

// require Aura.SqlQuery autoloader
// https://github.com/auraphp/Aura.SqlQuery
use Aura\SqlQuery\QueryFactory;

// init query factory
$queryFactory = new QueryFactory('Mysql');

// template engine
$loader = new \Twig\Loader\FilesystemLoader(__DIR__ . '/template');
$twig = new \Twig\Environment($loader);

// cache in debug mode! Stopped.
$twig = new \Twig\Environment($loader, [
    'cache' => __DIR__ .'/cache', 'debug'=>true,
]);

$action = (isset($_GET['action']) ? $_GET['action'] : null);
switch ($action) {
    case 'message_add':

/*
      'tags' => string 'ацууца' (length=12)
      'checkRules' => string 'on' (length=2)*/

        $username = (isset($_POST['username']) ? $_POST['username'] : "");
        $email = (isset($_POST['email']) ? urldecode($_POST['email']) : "");
        $message = (isset($_POST['message']) ? strip_tags(urldecode($_POST['message'])) : "");
        $homepage = (isset($_POST['homepage']) ? urldecode($_POST['homepage']) : "");
        $tags = (isset($_POST['tags']) ? urldecode($_POST['tags']) : "");

        if($username == "" || $email == "" || $message == "") {
            // TODO Ajax json
            echo "data_required";
            return;
        } else if(!(bool)preg_match('/^[A-z0-9]{1,}$/', $username) || (strlen($username) < 8 || strlen($username) > 32)) {
            echo "invalid_username";
            return;
        } else if(!(bool)preg_match('/^(([^<>()[\]\\.,;:\s@"]+(\.[^<>()[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/', $email) || strlen($email) > 64) {
            echo "invalid_email";
            return;
        } else if(strlen($message) < 10 || strlen($message) > 2048) {
            echo "invalid_message";
            return;
        } else if(strlen($homepage) > 128) {
            echo "invalid_homepage";
            return;
        } else if(strlen($tags) > 256) {
            echo "invalid_tags";
            return;
        }

        // PDO connection
        $pdo = new PDOconn();
        // init query factory
        $queryFactory = new QueryFactory('Mysql');
        // prepare insert transaction
        $insert = $queryFactory->newInsert();
        // insert into this table
        $insert->into('D_GUESTBOOK');
        $insert
            ->into('D_GUESTBOOK')      // INTO this table
            ->cols([                        // bind values as "(col) VALUES (:col)"
                'username' => $username,
                'email' => $email,
                'homepage' => $homepage,
                'text' => $message,
                'tags' => $tags,
                'create_date' => date_create()->format('Y-m-d H:i:s')
            ]);
        // prepare transaction and execute all rows
        $sth = $pdo->prepare($insert->getStatement());
        $sth->execute($insert->getBindValues());


        //var_dump($_POST);



        echo "success";
        break;

    default:

        // PDO connection
        $pdo = new PDOconn();

        /*
        $select = $queryFactory->newSelect();
        $select->cols([
            'id',
            'username',
            'email',
            'homepage',
            'text',
            'tags',
            'create_date',
            '@row_number:=@row_number+1 AS row_number'
        ])->from('D_GUESTBOOK') ;

        // prepare the statment
        $sth = $pdo->prepare($select->getStatement());
        // bind the values and execute
        $sth->execute($select->getBindValues());
        // get the results back as an associative array
        $result = $sth->fetchAll(PDO::FETCH_ASSOC);
        */

        $stmt = $pdo->query("SELECT id, username, email, homepage, text, tags, create_date, @row_number:=@row_number+1 AS row_num FROM `D_GUESTBOOK`, (SELECT @row_number:=0) AS t ORDER BY create_date DESC limit 100");
        $messages = $stmt->fetchAll();


        echo $twig->render('main.html', [
            'message' => true,
            'message_add' => true,
            'messages' => (isset($messages) ? $messages : null)
        ]);

        break;
}



template/js/main.js
document.onreadystatechange = function() {

    $(document).ready(function(){

      // Generate test data. Form submit
      $('#frmDemoRows').submit(function(e){
        e.preventDefault();
        e.stopImmediatePropagation();
        var inputRowCnt = $('#rows_cnt'),
            blockStatus = $('#blockStatus'),
            iconStatus = $('#iconStatus'),
            createBtn1 = $('#createBtn1');
        $.ajax({url:"install.php?action=generate",
          type: 'POST',
          data:"p1=" + inputRowCnt.val(),
          beforeSend:function(){
            blockStatus.html('')
            iconStatus.css("display","block")
            createBtn1.prop('disabled', true);
          },
          complete:function(){
            iconStatus.css("display","none")
            createBtn1.prop('disabled', false);
          },
          success:function(result) {
            if (result == 0) {
              blockStatus.html('<span style="color: red;">Необходимо значение от 10 до 10000</span>')
            } else {
              blockStatus.html('<span style="color: green;">' + result + '</span>')
            }
          },
          error:function(e){
            blockStatus.html('<span style="color: red;">' + e + '</span>')
          }
        })
        return false;
      });

      // Add comment. Form submit
      $('#frmAddMsg').submit(function(e){
        e.preventDefault();
        e.stopImmediatePropagation();
        var form    = $(this),
            btnSend = $('#btnMsgSend'),
            statusBlock = $('#blockStatus'),
            statusMsg1 = $('#statusMsg1'),
            statusMsg2 = $('#statusMsg2'),
            iconStatus    = $('#icon_status');
        $.ajax({url:"index.php?action=message_add",
          type: 'POST',
          data:form.serialize(),
          beforeSend:function(){
            statusBlock.css("display","none")
            statusMsg1.html('')
            statusMsg2.html('')
            iconStatus.css("display","block")
            btnSend.prop('disabled', true)
          },
          complete:function(){
            iconStatus.css("display","none")
            btnSend.prop('disabled', false)
          },
          success:function(result) {
            if (result=='success') {
              statusBlock.css("display", "block")
              statusMsg1.html('<h4 class="alert-heading">Успех!</h4>')
              statusMsg2.html('Ваше сообщение отправлено')
              $('#message').val('');
              $('#tags').val('');
            } else if (result=='data_required') {
              statusBlock.css("display", "block")
              statusMsg1.html('<h4 class="alert-heading">Ошибка!</h4>')
              statusMsg2.html('Проверьте введённые данные и повторите ещё раз')
            } else if (result=='invalid_username') {
              statusBlock.css("display", "block")
              statusMsg1.html('<h4 class="alert-heading">Ошибка!</h4>')
              statusMsg2.html('Неверное имя пользователя. Только латинские символы и цифры, от 8 символов до 32 символов.')
            } else if (result=='invalid_email') {
              statusBlock.css("display", "block")
              statusMsg1.html('<h4 class="alert-heading">Ошибка!</h4>')
              statusMsg2.html('Неверный E-Mail.')
            } else if (result=='invalid_message') {
              statusBlock.css("display", "block")
              statusMsg1.html('<h4 class="alert-heading">Ошибка!</h4>')
              statusMsg2.html('Сообщение должно содержать от 10 до 2048 символов.')
            } else if (result=='invalid_homepage') {
              statusBlock.css("display", "block")
              statusMsg1.html('<h4 class="alert-heading">Ошибка!</h4>')
              statusMsg2.html('Домашняя страница не может превышать 128 символов.')
            } else if (result=='invalid_tags') {
              statusBlock.css("display", "block")
              statusMsg1.html('<h4 class="alert-heading">Ошибка!</h4>')
              statusMsg2.html('Длина тегов не может превышать 256 символов.')
            } else {
              statusBlock.css("display","block")
              statusMsg1.html('<h4 class="alert-heading">Ошибка!</h4>')
              statusMsg2.html('Невозможно получить данные от сервера ')
            }
          },
          error:function(e){
            statusBlock.css("display","block")
            statusMsg1.html('Ошибка!')
            statusMsg2.html(e)
            console.log(e)
          }
        });
        return false;
      })

/*
      $('#frmAddMsg').on('submit', function(e){
        var form    = $(this),
            input   = $('.check-recover-email'),
            span    = $('.recover-error'),
            email   = input.val();
        span.text('');
        e.preventDefault(); // <=================== Here
        $.ajax({
          url: 'ajax/check-email',
          type: 'POST',
          data: form.serialize(),
          success: function(response){
            if ( response == 0 ) {
              // ============================ Not here, this would be too late
              span.text('email does not exist');
            }
          }
        });
      });
*/

    });



}



install.php
<?php

// maximum message length (minimum = MESSAGE_MAX_LENGTH / 6)
define("MESSAGE_MAX_LENGTH", 1024);
// maximum keywords (minimum = KEYWORDS_LIMIT / 2)
define("KEYWORDS_LIMIT", 15);
// minimum keyword length
define("KEYWORD_MIN_LENGTH", 7);

require_once __DIR__ . '/vendor/autoload.php';
require_once __DIR__ . '/functions/PDOconn.php';

// require the Faker autoloader
// https://github.com/fzaninotto/faker
use Faker\Factory;
//require_once 'vendor/fzaninotto/faker/src/autoload.php';

// require Aura.SqlQuery autoloader
// https://github.com/auraphp/Aura.SqlQuery
use Aura\SqlQuery\QueryFactory;

// template engine
$loader = new \Twig\Loader\FilesystemLoader(__DIR__ . '/template');
$twig = new \Twig\Environment($loader);

// cache in debug mode! Stopped.
$twig = new \Twig\Environment($loader, [
    'cache' => __DIR__ .'/cache', 'debug'=>true,
]);
$response = null;

$action = isset($_GET['action']);
switch ($action) {

    case 'generate':

        if(!isset($_POST['p1']) || !($_POST['p1'] >= 10 && $_POST['p1'] <= 10000)) {
            // TODO Ajax json
            //echo "";
            return;
        }
        $cntTestRows = $_POST['p1'];

        createTables();

        // create a Faker instance
        $faker = Faker\Factory::create('ru_RU');

        // PDO connection
        $pdo = new PDOconn();
        // init query factory
        $queryFactory = new QueryFactory('Mysql');

        // prepare insert transaction
        $insert = $queryFactory->newInsert();
        $htmlTestRows = "";
        $goodRows = 0;
        $badRows = 0;
        $time_start = microtime(true);

        for ($i = 0; $i < $cntTestRows; $i++) {
            $arFaker = createFakeData($faker);
            if($arFaker != null){
                $arFaker['keywords'] = "";
                $keywordsCnt = 0;
                do {
                    $keyword = $arFaker['words'][rand(0, count($arFaker['words'])-1)];
                    if(mb_strlen($keyword, 'UTF-8') >= KEYWORD_MIN_LENGTH) {
                        $arFaker['keywords'] .= $keyword . ", ";
                        $keywordsCnt++;
                    }
                } while ($keywordsCnt < rand(round(KEYWORDS_LIMIT / 2), KEYWORDS_LIMIT));
                $arFaker['keywords'] = substr($arFaker['keywords'], 0, -2);
                // random created date and time
                $created = $faker->dateTimeThisYear($max = 'now', $timezone = null);//->format('d.m.Y - H:m:s');

                $htmlTestRows .= "<p>" . PHP_EOL;
                $htmlTestRows .= "<b>Username:</b> " . $arFaker['username'] . "<br/>" . PHP_EOL;
                $htmlTestRows .= "<b>E-Mail:</b> " . $arFaker['address'] . "<br/>" . PHP_EOL;
                $htmlTestRows .= "<b>Homepage:</b> " . $arFaker['homepage'] . "<br/>" . PHP_EOL;
                $htmlTestRows .= "<b>Message:</b> " . $arFaker['message'] . "<br/>" . PHP_EOL;
                $htmlTestRows .= "<b>Tags:</b> " . $arFaker['keywords'] . "<br/>" . PHP_EOL;
                $htmlTestRows .= "<b>Created:</b> " . $created->format('d.m.Y - G:i:s') . "<br/>" . PHP_EOL;
                $htmlTestRows .= "<p/>" . PHP_EOL;

                // insert into this table
                $insert->into('D_GUESTBOOK');
                $insert
                    ->into('D_GUESTBOOK')      // INTO this table
                    ->cols([                        // bind values as "(col) VALUES (:col)"
                        'username' => $arFaker['username'],
                        'email' => $arFaker['address'],
                        'homepage' => $arFaker['homepage'],
                        'text' => $arFaker['message'],
                        'tags' => $arFaker['keywords'],
                        'create_date' => $created->format('Y-m-d G:i:s')
                    ]);
                // set up Multiple rows ...
                //$insert->addRow();
                // prepare transaction and execute all rows
                $sth = $pdo->prepare($insert->getStatement());
                $sth->execute($insert->getBindValues());
                $goodRows++;
            } else {
                $badRows++;
                // 1000 errors its very bad... infinity loop? Nope. Break.
                if($badRows >= 1000) {
                    break;
                }
            }
        }


        //$sth->rowCount()

        $response = "Генерация тестовых данных завершена за " . number_format((float)microtime(true) - $_SERVER["REQUEST_TIME_FLOAT"], 2, '.', '') . " сек. Количество добавленных записей: " . $goodRows . ", ошибочных генераций: " . $badRows;
        echo $response;
        //'fa-check-circle-o'
        break;

    default:

        $template = $twig->load('main.html');
        echo $template->render([
            'demo' => true
        ]);

}

/*$select = $queryFactory->newSelect();
$select->cols([
    count('id')
])->from('D_GUESTBOOK') ;

// prepare the statment
$sth = $pdo->prepare($select->getStatement());
// bind the values and execute
$sth->execute($select->getBindValues());
// get the results back as an associative array
$result = $sth->fetch(PDO::FETCH_ASSOC);
var_dump($result);*/



function createFakeData($faker) {
    $arFaker = null;
    // random name
    $arFaker['username'] = $faker->word . $faker->word . $faker->numberBetween($min = 1954, $max = 2001);
    // random address
    $arFaker['address'] = $faker->freeEmail;
    // random Homepage
    $arFaker['homepage'] = $faker->optional($weight = 0.4)->domainName;
    // random text
    $arFaker['message'] = $faker->realText($maxNbChars = rand(round(MESSAGE_MAX_LENGTH / 6), MESSAGE_MAX_LENGTH));
    // random tags, with utf-8
    $wordsPattern = "~[^\p{L}\\'\-\\xC2\\xAD]+~u";
    $arFaker['words'] = preg_split($wordsPattern, $arFaker['message'], -1, PREG_SPLIT_NO_EMPTY);
    if(is_null($arFaker['message']) || strlen($arFaker['message']) < 10 || is_null($arFaker['username']) || is_null($arFaker['address'])){
        return null;
    }
    return $arFaker;
}

function createTables() {
    $pdo = new PDOconn();

    $sqlCreateTableGuestBook = <<<SQL
    DROP TABLE IF EXISTS `D_GUESTBOOK`;
    
    SET SQL_MODE = "NO_AUTO_VALUE_ON_ZERO";
    SET time_zone = "+00:00";
    
    /*!40101 SET @OLD_CHARACTER_SET_CLIENT=@@CHARACTER_SET_CLIENT */;
    /*!40101 SET @OLD_CHARACTER_SET_RESULTS=@@CHARACTER_SET_RESULTS */;
    /*!40101 SET @OLD_COLLATION_CONNECTION=@@COLLATION_CONNECTION */;
    /*!40101 SET NAMES utf8 */;
    
    --
    -- Структура таблицы `D_GUESTBOOK`
    --
    
    CREATE TABLE IF NOT EXISTS `D_GUESTBOOK` (
    `id` int(11) NOT NULL,
      `user_id` int(11) DEFAULT NULL COMMENT 'When User schon registriert ist',
      `username` varchar(64) NOT NULL DEFAULT 'NULL' COMMENT 'Anmeldename des Benutzers',
      `email` varchar(64) NOT NULL DEFAULT 'NULL' COMMENT 'E-Mail-Adresse des Kontos',
      `homepage` varchar(128) DEFAULT NULL COMMENT 'Web-site des Benutzers',
      `text` varchar(2048) NOT NULL DEFAULT 'NULL' COMMENT 'Nachricht, die Benutzer eingegeben',
      `tags` varchar(256) DEFAULT NULL COMMENT 'die Stichwörter',
      `create_date` timestamp NOT NULL DEFAULT CURRENT_TIMESTAMP COMMENT 'Datum, wann die Nachricht hinzugefügt',
      `is_approved` int(1) NOT NULL DEFAULT '1' COMMENT '1 - When Daten akzeptiert sind'
    ) ENGINE=InnoDB DEFAULT CHARSET=utf8 COMMENT='The Guestbook';
    
    --
    -- Индексы сохранённых таблиц
    --
    
    --
    -- Индексы таблицы `D_GUESTBOOK`
    --
    ALTER TABLE `D_GUESTBOOK`
     ADD PRIMARY KEY (`id`), ADD KEY `user_id` (`user_id`);
    
    --
    -- AUTO_INCREMENT для сохранённых таблиц
    --
    
    --
    -- AUTO_INCREMENT для таблицы `D_GUESTBOOK`
    --
    ALTER TABLE `D_GUESTBOOK`
    MODIFY `id` int(11) NOT NULL AUTO_INCREMENT;
    --
    -- Ограничения внешнего ключа сохраненных таблиц
    --
    
    --
    -- Ограничения внешнего ключа таблицы `D_GUESTBOOK`
    --
    ALTER TABLE `D_GUESTBOOK`
    ADD CONSTRAINT `D_GUESTBOOK_ibfk_1` FOREIGN KEY (`user_id`) REFERENCES `S_USERS` (`id`);
    
    /*!40101 SET CHARACTER_SET_CLIENT=@OLD_CHARACTER_SET_CLIENT */;
    /*!40101 SET CHARACTER_SET_RESULTS=@OLD_CHARACTER_SET_RESULTS */;
    /*!40101 SET COLLATION_CONNECTION=@OLD_COLLATION_CONNECTION */;
    SQL;

    $crtResult = $pdo->query($sqlCreateTableGuestBook);
    //var_dump($crtResult);
}